FRs #1055/1073: Refactor Provider Identification, Host Provider, Invoice Issuer (#1139)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: specification/datasets/cost_and_usage/columns/hostprovidername.md

Reviewer: Rather than add a series of revisions here, I have instead opted to issue a recommend change to Host Provider Name via PR #1524.  In my opinion, it simplifies the normative requirements, and it gives CSPs the option not to publish it.  @rileyjenk Please have a look at that and lmk what you think.

Reviewer: Per discussion in Oct 13 Maintainers, I have revised PR #1524 to revise Host Provider.  @rileyjenk please give it a look and merge in if/when you think it looks good.

# Host Provider

A Host Provider is an entity that provides the underlying infrastructure on which the [*resources*](#glossary:resource) or [*services*](#glossary:service) of one or more Service Providers are deployed.

The HostProvider column adheres to the following requirements:

* The HostProvider column MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset) and represents the service provider where [*resources*](#glossary:resource) or [*services*](#glossary:service) are deployed on.
* For Service Providers that allow their customer to select or know the host provider where resources or services are deployed, this SHOULD be the name of that selected or visible host provider

> * For Service Providers that DO NOT allow their customer to select or know the host provider where resources or services are deployed, this SHOULD be the name of the [Service Provider](#serviceprovider)

Author: Need to add for the case where there is no host_provider known.

# Host Provider Name

Host Provider Name is the name of the entity that provides the underlying infrastructure on which the [*resources*](#glossary:resource) or [*services*](#glossary:service) of one or more Service Providers are deployed.

The HostProviderName column adheres to the following requirements:

> * The HostProviderName column MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset) and represents the service provider where [*resources*](#glossary:resource) or [*services*](#glossary:service) are deployed on.
> * For Service Providers that allow their customer to select or know the host provider where resources or services are deployed, this SHOULD be the name of that selected or visible host provider
> * For Service Providers that DO NOT allow their customer to select or know the host provider where resources or services are deployed, this SHOULD be the name of the [Service Provider](#serviceprovider)
> * For charges that are not relevant to a hosting scenario, such as professional service charges, license purchase, taxes, or refunds, this MAY be "NOT APPLICABLE."
> * This column MUST be of type String and MAY contain null values when a line item does not pertain to a resource or services with a specific host provider, such as purchases, refunds, or taxes.

Reviewer: @rileyjenk I have taken a stab at rewriting these normative requirements following our guidelines.  I have not attempted to change them materially; merely tried to get them into the right format.  @ijurica, please have a look as well.

```suggestion
* HostProviderName MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset).
* HostProviderName MUST be of type String.
* HostProviderName MUST conform to [StringHandling](#stringhandling) requirements.
* HostProviderName MUST NOT be null when a *charge* pertains to *resources* or *services*. 
* HostProviderName values are defined as follows:
  * HostProviderName SHOULD reflect the name of the [*host provider*](#glossary:host-provider) of deployed *resources* or *services* when this information is available to customers.
  * HostProviderName MAY be "Not Applicable" for *charges* that are not relevant to a hosting scenario (e.g., professional services, licenses, taxes, refunds).
  * HostProviderName MUST equal [ServiceProviderName](#serviceprovidername) in all other cases.
```

# Host Provider Name

Host Provider Name is the name of the entity that provides the underlying infrastructure on which the [*resources*](#glossary:resource) or [*services*](#glossary:service) of one or more Service Providers are deployed.

The HostProviderName column adheres to the following requirements:

* HostProviderName MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset).
* HostProviderName MUST be of type String.
* HostProviderName MUST conform to [StringHandling](#stringhandling) requirements.

> * HostProviderName MUST NOT be null when a *charge* pertains to *resources* or *services*. 

Reviewer 2: I suggest we standardize on "Not Applicable" and avoid using null.

```suggestion
* HostProviderName MUST NOT be null. 
```

> * HostProviderName MUST NOT be null when a *charge* pertains to *resources* or *services*. 

Reviewer 2: linter fix.

```suggestion
* HostProviderName MUST NOT be null when a *charge* pertains to *resources* or *services*.
```

Reviewer: I addressed in a separate linter commit just now.  Resolving.

* HostProviderName values are defined as follows:
  * HostProviderName SHOULD reflect the name of the [*host provider*](#glossary:host-provider) of deployed *resources* or *services* when this information is available to customers.
  * HostProviderName MAY be "Not Applicable" for *charges* that are not relevant to a hosting scenario (e.g., professional services, licenses, taxes, refunds).

>   * HostProviderName MUST equal [ServiceProviderName](#serviceprovidername) in all other cases.

Reviewer 2: - I suggest we stick either to null or "Not applicable". (Reminder: In the examples we've used "N/A").
- I believe there are use cases in which we MUST use "Not Applicable".

```suggestion
* HostProviderName MUST NOT be null.
* HostProviderName values are defined as follows:
  * HostProviderName MUST reflect the name of the host provider when explicitly selected by the customer.
  * HostProviderName MUST be "Not Applicable" when the associated [ServiceName](#servicename) does not involve deployment on any underlying infrastructure (e.g., professional services, software licenses).
  * HostProviderName MAY equal [ServiceProviderName](#serviceprovidername) when the service provider does not provide visibility into the underlying hosting provider.
  * HostProviderName MAY be "Not Applicable" when the information about the entity providing the underlying infrastructure cannot be uniquely determined (e.g., when the [ChargeCategory](#chargecategory) is "Tax" or "Adjustment").
```

# Host Provider Name

> Host Provider Name is the name of the entity that provides the underlying infrastructure on which the [*resources*](#glossary:resource) or [*services*](#glossary:service) of one or more Service Providers are deployed.

Reviewer 3: What is the final part of the sentence, 'one or more service providers are deployed', relevant here? Generally, we talk about the column in the context of the row.. so wouldn't this just be fine without that final part? We also typically give an example of what this column is useful for 'e.g. This column is commonly used for identifying the entity that ____ for use case ___'

Author: Good call out. I can fix this.

The HostProviderName column adheres to the following requirements:

> * HostProviderName MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset).

Reviewer: I do not think it correct to require CSPs to populate this column.  They should be allowed to skip it if it will always be fully duplicative with Service Provider.

```suggestion
* HostProviderName MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset) when a data generator publicly delineates between a service provider and a host provider.
```

Reviewer: Resolving in favor of PR #1524.

* HostProviderName MUST be of type String.
* HostProviderName MUST conform to [StringHandling](#stringhandling) requirements.

> * HostProviderName values are defined as follows:

Reviewer 2: According to the guidelines, a nullability requirement is required. 

We should probably specify the nullability more precisely, but at this point, I believe this is the minimum:

```suggestion
* HostProviderName MAY be null.
* HostProviderName values are defined as follows:
```

Reviewer: Nullability is mentioned farther down.  That said, we have more than that to address.  I have taken a stab via PR #1524.

  * HostProviderName MUST reflect the name of the host provider when explicitly selected by the customer.

>   * HostProviderName MUST reflect the name of the host provider when the service provider does not provide visibility into the underlying hosting provider.

Reviewer 2: ```suggestion
  * HostProviderName MUST reflect the name of the host provider when the service provider does provide visibility into the underlying hosting provider.
```

Reviewer: I believe we need to address this section more holistically.  I have proposed something via PR #1524.

>   * HostProviderName MUST be "Not Applicable" when the associated [ServiceName](#servicename) does not involve deployment on any underlying infrastructure (e.g., professional services, software licenses).

Reviewer 3: Sorry, but is this a new thing? Generally, we've said we want to use null and not use 'placeholder' values - which was a requirement (not sure if its still in place)

Author: Check this out now. Updated the conditionals.

  * HostProviderName MAY equal [ServiceProviderName](#serviceprovidername) when the service provider does not provide visibility into the underlying hosting provider.
  * HostProviderName MAY be "Not Applicable" when the information about the entity providing the underlying infrastructure cannot be uniquely determined (e.g., when the [ChargeCategory](#chargecategory) is "Tax" or "Adjustment").

> See [Appendix: Entity Identification Examples](#entityidentification) section for examples of Host Provider values for various use case scenarios.

Reviewer 2: ```suggestion
See [Appendix: Participating Entity Identification Examples](#participatingentityidentificationexamples) section for examples of Host Provider values across various use case scenarios.
```
